tsdate/util.py

```diff
@@ -121,7 +124,104 @@ def preprocess_ts(tree_sequence, minimum_gap=1000000, trim_telomeres=True):
+        mutation associated with a site. If "arithmetic", return the age of the oldest
+        mutation associated with each site, calculated using the arithmetic mean of
+        the parent and child nodes of the mutation. If "geometric", use the geometric
+        mean. "parent" returns the age of the node of the parent.
+        Default: "child"
+    :param bool ignore_multiallelic: If True, return age of oldest mutation associated
+        with an allele at multiallelic sites. If False, return tskit.UNKNOWN_TIME at any
+        multiallelic sites. (defaults to True).
+        Default: True
+    :return: An array of length tree_sequence.num_sites giving the estimated site time
+        of each site.
+    :rtype numpy.array
+
+    """
+    if tree_sequence.num_sites < 1:
+        raise ValueError("Invalid tree sequence: no sites present")
+    if mutation_age not in ["arithmetic", "geometric", "child"]:
+        raise ValueError(
+            "mutation_age parameter must be 'arithmetic', 'geometric' or 'child'")
+    # Epsilon value to date sites with > 1 mutation where all mutations are singletons
+    eps = 1e-6
+    sites_time = np.empty(tree_sequence.num_sites)
+    sites_time[:] = -np.inf
+    nodes_age = nodes_time(tree_sequence, unconstrained=unconstrained)
+
+    for tree in tree_sequence.trees():
+        for site in tree.sites():
+            alleles = set([site.ancestral_state])
+            for mutation in site.mutations:
+                alleles.add(mutation.derived_state)
+                if mutation_age == "child":
+                    age = nodes_age[mutation.node]
+                else:
+                    parent_node = tree.parent(mutation.node)
+                    if mutation_age == "arithmetic":
+                        age = (nodes_age[mutation.node] + nodes_age[parent_node]) / 2
+                    elif mutation_age == "geometric":
+                        age = np.sqrt(nodes_age[mutation.node] * nodes_age[parent_node])
+                if sites_time[site.id] < age:
+                    sites_time[site.id] = age
+            if len(site.mutations) > 1 and sites_time[site.id] == 0:
+                sites_time[site.id] = eps
+            if ignore_multiallelic is False:
+                if len(alleles) > 2:
+                    sites_time[site.id] = tskit.UNKNOWN_TIME
+    return sites_time
+
+
+def add_sampledata_times(samples, sites_time):
+    """
+    Return a tsinfer.SampleData file with the given times associated with sites. Also
+    checks that each sites' time is at least as old as the oldest historic sample
+    carrying a derived allele at that site.
+
+    :param tsinfer.formats.SampleData samples: A tsinfer SampleData object to
+        add site times to. If both a SampleData file is provided and constrain_historic
+        is True, use any historic individuals in this SampleData file to constrain site
+        times.
+    :return A tsinfer.SampleData file with sites_times
```

Comment:
Don't all SampleData feels have sites_times? So this needs a little more explanation